### PR TITLE
Cleanup requires_asap()

### PR DESCRIPTION
**Pull request by grantisu**
The biggest change here is that a server will see exceptions from failed key lookups.

- Allow server to handle server errors
- Raise more consistent exceptions
- Perform some minor code cleanup

### CODE REVIEW

File: atlassian_jwt_auth/key.py [hunk 1]
```diff
@@ -93,7 +93,10 @@ def retrieve(self, key_identifier, **requests_kwargs):
             key_identifier = KeyIdentifier(key_identifier)
 
         url = self.base_url + key_identifier.key_id
-        return self._retrieve(url, requests_kwargs)
```
Comment (dbaxa) [inline atlassian_jwt_auth/key.py]: I prefer not to wrap exceptions like this unless really needed.

File: atlassian_jwt_auth/contrib/aiohttp/key.py [hunk 2]
```diff
@@ -21,8 +22,16 @@ def _get_session(self):
...
+                                           **requests_kwargs)
+        except aiohttp.ClientError as e:
+            raise requests.RequestException(e)
+
+        try:
+            resp.raise_for_status()
+            self._check_content_type(url, resp.headers['content-type'])
+            return await resp.text()
+        except aiohttp.http.HttpProcessingError as e:
+            raise requests.HTTPError(e)
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/aiohttp/key.py]: Why wrap this exception?

File: atlassian_jwt_auth/key.py (hunk 1 above)
Comment (grantisu) [reply atlassian_jwt_auth/key.py]: That would mean that any code that cares about handling failures to retrieve public keys must know all the details about how those keys are received, which is a leaky abstraction.

I'm fine with a more elaborate wrapping scheme than what's currently here, e.g. adding an `original_exception` attribute to `PublicKeyRetrieverException`.

File: atlassian_jwt_auth/contrib/aiohttp/key.py (hunk 2 above)
Comment (grantisu) [reply atlassian_jwt_auth/contrib/aiohttp/key.py]: None of the code that catches key retrieval exceptions or request exceptions handles aiohttp exceptions. Normalizing to one type of exception fixes those bugs and makes for easier code reuse.

I'm not set on `requests.HTTPError` here: we could use something else or create a new exception class.

File: atlassian_jwt_auth/contrib/aiohttp/key.py [hunk 3]
```diff
@@ -21,8 +22,16 @@ def _get_session(self):
         return aiohttp.ClientSession(loop=self.loop)
 
     async def _retrieve(self, url, requests_kwargs):
```
Comment (erikvanzijst) [inline atlassian_jwt_auth/contrib/aiohttp/key.py]: Unrelated, but is this library meant to support Python 2?

Edit: never mind, I now see how this is used only on p3k.

File: atlassian_jwt_auth/contrib/aiohttp/key.py (hunk 2 above)
Comment (erikvanzijst) [reply atlassian_jwt_auth/contrib/aiohttp/key.py]: Good catch!

File: atlassian_jwt_auth/contrib/django/decorators.py [hunk 4]
```diff
@@ -30,43 +28,29 @@ def requires_asap_wrapper(request, *args, **kwargs):
...
-                message = 'Unauthorized: Invalid token'
-                exception = e
-            if message is not None:
-                logger = logging.getLogger(__name__)
-                logger.error(message,
-                             extra={'original_message': str(exception)})
-
+                # Invalid token
+                message = 'Unauthorized: %s' % e
                 return HttpResponse(message, status=401)
```
Comment (erikvanzijst) [inline atlassian_jwt_auth/contrib/django/decorators.py]: Shouldn't we be including a WWW-authenticate response header as per [RFC-6750](https://tools.ietf.org/html/rfc6750#section-3), considering we are using Bearer tokens?

File: atlassian_jwt_auth/key.py (hunk 1 above)
Comment (erikvanzijst) [reply atlassian_jwt_auth/key.py]: I agree with Grant. This code here should not be coupled to the fact that there is a aiohttp retriever.
Currently the retriever interface exposes `requests` and that's what clients should anticipate. While we *could* abstract the `RequestException` with something implementation-agnostic, I think it wouldn't add too much value.

File: atlassian_jwt_auth/key.py [hunk 5]
```diff
@@ -29,11 +29,11 @@ class KeyIdentifier(object):
     """ This class represents a key identifier """
 
     def __init__(self, identifier):
-        self.__key_id = validate_key_identifier(identifier)
+        self._key_id = validate_key_identifier(identifier)
 
     @property
     def key_id(self):
-        return self.__key_id
+        return self._key_id
```
Comment (dbaxa) [inline atlassian_jwt_auth/key.py]: This change seems unrelated to this pull request.

File: atlassian_jwt_auth/contrib/tests/django/test_django.py [hunk 6]
```diff
@@ -88,16 +88,16 @@ def test_request_with_invalid_audience_is_rejected(self):
             response = self.client.get(reverse('expected'),
                                        HTTP_AUTHORIZATION=b'Bearer ' + token)
 
-        self.assertContains(response, 'Unauthorized: Invalid token',
-                            status_code=401)
+        self.assertContains(response, 'Forbidden',
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/tests/django/test_django.py]: I prefer the previous response message because it provides details on why the request failed. Also, switching from `401` to `403` may seem useful but there seems to be a split about what to do in the case of an authorisation failure amongst asap libraries. For example, the go-asap library uses 401. Unless we have a good reason to change this I am inclined to say stick with `401`.

File: atlassian_jwt_auth/contrib/tests/django/test_django.py [hunk 7]
```diff
@@ -88,16 +88,16 @@ def test_request_with_invalid_audience_is_rejected(self):
...
+                            status_code=403)
 
     def test_request_with_invalid_token_is_rejected(self):
         with override_settings(**self.test_settings):
             response = self.client.get(
                 reverse('expected'),
                 HTTP_AUTHORIZATION=b'Bearer notavalidtoken')
 
-        self.assertContains(response, 'Unauthorized: Invalid token',
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/tests/django/test_django.py]: Why change this?

File: atlassian_jwt_auth/contrib/tests/django/test_django.py [hunk 8]
```diff
@@ -119,7 +119,8 @@ def test_request_with_invalid_issuer_is_rejected(self):
             response = self.client.get(reverse('expected'),
                                        HTTP_AUTHORIZATION=b'Bearer ' + token)
 
-        self.assertContains(response, 'Unauthorized: Invalid token issuer',
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/tests/django/test_django.py]: This doesn't really need to change but I don't mind it.

File: atlassian_jwt_auth/contrib/aiohttp/key.py (hunk 2 above)
Comment (dbaxa) [reply atlassian_jwt_auth/contrib/aiohttp/key.py]: I am not going to agree on this change because you will make it harder for users of the `aiohttp` code to obtain or catch a specific exception that might occur. 

Importantly, 
* `aiohttp.ClientError` extends `Exception` and does not have a `request` nor a `response` attribute.
* `aiohttp.HttpProcessingError` extends `Exception` and has different attributes than `requests.HTTPError`.

File: atlassian_jwt_auth/contrib/django/decorators.py [hunk 9]
```diff
@@ -30,43 +28,29 @@ def requires_asap_wrapper(request, *args, **kwargs):
...
-                # Error parsing or getting public key
-                message = 'Unauthorized: Unable to retrieve public key'
-                exception = e
-            except InvalidIssuerError as e:
-                message = 'Unauthorized: Invalid token issuer'
-                exception = e
+            except InvalidAudienceError as e:
+                # Audience mismatch
+                message = 'Forbidden: %s' % e
+                return HttpResponse(message, status=403)
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/django/decorators.py]: I prefer the previous response message because it provides details on why the request failed. Also, switching from `401` to `403` may seem useful but there seems to be a split about what to do in the case of an authorisation failure amongst asap libraries. For example, the go-asap library uses 401. Unless we have a good reason to change this I am inclined to say stick with `401`.

File: atlassian_jwt_auth/key.py (hunk 5 above)
Comment (grantisu) [reply atlassian_jwt_auth/key.py]: Sorta. This PR started as some minor cleanup, then spiraled out a bit.

I can pull the minor parts out if they're bugging anybody.

File: atlassian_jwt_auth/contrib/tests/django/test_django.py (hunk 6 above)
Comment (grantisu) [reply atlassian_jwt_auth/contrib/tests/django/test_django.py]: After re-reading [the spec](https://s2sauth.bitbucket.io/spec/), I think you're right: an invalid audience should result in a `401`, since it kinda makes sense that one server shouldn't recognize credentials that are only valid on another server.

But then we should probably make invalid issuers a `403`. A perfectly valid token from a non-whitelisted issuer/user sounds like an authorization error. Also, in the absence of a subject the issuer is the subject, and rejecting a (properly authenticated) request based on the subject feels like an authorization error.

I'll back out the status code changes in this PR.

File: atlassian_jwt_auth/contrib/aiohttp/key.py (hunk 2 above)
Comment (dbaxa) [reply atlassian_jwt_auth/contrib/aiohttp/key.py]: > If different callers need to have different exceptions, then they probably need to have a custom interface as well.

Right so just to clarify I wasn't thinking that a `django` or` flask` application would use the <django/flask> contrib code with an `asyncio` `HTTPSPublicKeyRetriever`. Perhaps you are thinking of this?

Comment (grantisu) [reply atlassian_jwt_auth/contrib/aiohttp/key.py]: I was thinking that `HTTPSPublicKeyRetriever` shouldn't have to care what's calling it, so long as both sides follow the same interface.

File: atlassian_jwt_auth/contrib/server/helpers.py [hunk 10]
```diff
@@ -1,45 +1,40 @@
...
-        message = 'Unauthorized: Invalid token issuer'
-        exception = e
+        message = 'Forbidden: Invalid token issuer'
+        return build_response(message, status=403)
     except jwt.exceptions.InvalidTokenError as e:
         # Something went wrong with decoding the JWT
         message = 'Unauthorized: Invalid token'
-        exception = e
-    if message is not None:
-        logger = logging.getLogger(__name__)
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/server/helpers.py]: Why are we removing the logging? (I can moving logging to another location but it is imho useful and this change is just removing it).

File: atlassian_jwt_auth/contrib/server/helpers.py [hunk 11]
```diff
@@ -1,45 +1,40 @@
...
+        return build_response('Unauthorized', status=401, headers={
+                              'WWW-Authenticate': 'Bearer'})
+    verifier = get_verifier()
     try:
-        asap_claims = parse_jwt_func(verifier, auth[1])
+        asap_claims = parse_jwt_func(verifier, auth)
         if verify_issuers_func is not None:
             verify_issuers_func(asap_claims, issuers)
         asap_claim_holder.asap_claims = asap_claims
-    except requests.exceptions.HTTPError as e:
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/server/helpers.py]: Actually this can occur if the key cannot be found and that's not the server's fault at all.

File: atlassian_jwt_auth/contrib/server/helpers.py [hunk 12]
```diff
@@ -1,45 +1,40 @@
...
-        # Couldn't find key in key server
-        message = 'Unauthorized: Invalid key'
-        exception = e
-    except (requests.exceptions.ConnectionError,
-            PublicKeyRetrieverException) as e:
-        message = 'Unauthorized: Backend server connection error'
```
Comment (dbaxa) [inline atlassian_jwt_auth/contrib/server/helpers.py]: It may make sense to make this a non-4xx response.

**Comment by dbaxa** [comment]
I would like to cherry-pick 30dd4a3038dfb657c0d996d1762d2679e849c898 , d434113546ca27a45a34e583404a4f7638575068 6c9e2ed1763e221a9263a8599a942203d9bbfcc4 into a new pull request (with some minor changes).

File: atlassian_jwt_auth/contrib/server/helpers.py (hunk 10 above)
Comment (grantisu) [reply atlassian_jwt_auth/contrib/server/helpers.py]: Logging client errors tends to flood the logs with useless entries, especially authorization errors.

For example, a client that only sends authorization headers in response to a 401 will generate an extra "error" log for each request. These should already show up in the request logs, so dropping the logging here shouldn't drop any interesting information.

File: atlassian_jwt_auth/contrib/server/helpers.py (hunk 11 above)
Comment (grantisu) [reply atlassian_jwt_auth/contrib/server/helpers.py]: Good point.

This should probably be handled at the point where the request is made: any 4xx response from the key server should raise an `InvalidTokenError` (eventually resulting in a 401), while any other error (can't reach remote server, 5xx response, etc.) should be a `PublicKeyRetrieverException`.

Maybe `InvalidIssuerError` could represent a 4xx response from the key server, and something like `ForbiddenIssuerException` could be a valid issuer that isn't in the whitelist (with a 403 response)?

File: atlassian_jwt_auth/contrib/server/helpers.py (hunk 10 above)
Comment (dbaxa) [reply atlassian_jwt_auth/contrib/server/helpers.py]: @grantisu I'd prefer to change the log level (to say debug or info) then drop this logging.

**Comment by grantisu** [comment]
Remaining changes from this PR have been updated/moved to #65
